### PR TITLE
Iterate over all FO_ENUM values

### PR DESCRIPTION
**Pull request by o01eg**
Introduce function which can return iterable object with all values of FO_ENUM. Could be used to wrap values for python or future Godot binding.

### CODE REVIEW

File: default/python/AI/ProductionAI.py [hunk 1]
```diff
@@ -1080,7 +1080,7 @@ def generate_production_orders():
     building_name = "BLD_XENORESURRECTION_LAB"
     queued_xeno_lab_locs = [element.locationID for element in production_queue if element.name == building_name]
     for pid in state.get_all_empire_planets():
-        if pid in queued_xeno_lab_locs or not empire.canBuild(fo.buildType.building, building_name, pid):
+        if pid in queued_xeno_lab_locs or not empire.canBuild(fo.buildType.BT_BUILDING, building_name, pid):
```
Comment (geoffthemedio) [inline default/python/AI/ProductionAI.py:1083]: Don't need to update https://github.com/freeorion/freeorion/blob/master/default/python/freeorion.pyi#L1957 as well?

Comment (o01eg) [reply default/python/AI/ProductionAI.py:1083]: Yep, should be updated too.

Comment (o01eg) [reply default/python/AI/ProductionAI.py:1083]: Fixed.

**Comment by geoffthemedio** [comment]
server crashes when starting a game with this applied...

**Comment by o01eg** [comment]
@geoffthemedio Was some error in log or console?

**Comment by geoffthemedio** [comment]
Not really...
end of server log:
````
10:29:18.750948 {0x00005d6c} [info] server : dependencyversions.cpp:81 : Dependency versions from headers:
10:29:18.750948 {0x00005d6c} [info] server : dependencyversions.cpp:84 : Boost: 1_70
10:29:18.750948 {0x00005d6c} [info] server : dependencyversions.cpp:84 : Python: 3.5.7
10:29:18.750948 {0x00005d6c} [info] server : dependencyversions.cpp:84 : zlib: 1.2.11
10:29:18.750948 {0x00005d6c} [debug] server : commonframework.cpp:30 : Initializing FreeOrion Python interface
10:29:18.750948 {0x00005d6c} [debug] server : commonframework.cpp:41 : Python home set to C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt
10:29:18.750948 {0x00005d6c} [debug] server : commonframework.cpp:44 : Python program name set to C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt\freeoriond.exe
10:29:18.750948 {0x00005d6c} [debug] server : serverframework.cpp:53 : Initializing server Python imports
10:29:18.822964 {0x00005d6c} [debug] server : commonframework.cpp:57 : Python initialized
10:29:18.822964 {0x00005d6c} [debug] server : commonframework.cpp:58 : Python version: 3.5.7 (default, Mar  8 2020, 18:06:18) [MSC v.1916 32 bit (Intel)]
10:29:18.822964 {0x00005d6c} [debug] server : commonframework.cpp:59 : Python prefix: C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt
10:29:18.822964 {0x00005d6c} [debug] server : commonframework.cpp:60 : Python module search path: C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt\python35.zip;;C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt\Lib\python35;C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt\Lib\python35\lib-dynload;C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt\Lib\python35\lib-dynload\RelWithDebInfo;C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt\Lib\python35\plat-win32;C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt
10:29:18.822964 {0x00005d6c} [debug] server : commonframework.cpp:67 : Initializing C++ interfaces for Python
10:29:18.822964 {0x00005d6c} [debug] server : serverframework.cpp:60 : Initializing server Python modules
10:29:18.929988 {0x00005d6c} [info] log : logger.cpp:200 : Added logger named "python"
10:29:18.929988 {0x00005d6c} [info] log : logger.cpp:339 : Setting "python" logger threshold to "debug".
10:29:18.929988 {0x00005d6c} [debug] log : loggerwithoptionsdb.cpp:105 : Configure log source "python" from optionsDB using threshold debug
10:29:18.929988 {0x00005d6c} [debug] python : configure_logging.py:221 : Python stdout and stderr are redirected to ai process.
10:29:18.929988 {0x00005d6c} [info] python : configure_logging.py:242 : The python logger is initialized with a log level of DEBUG
````

end of client log:
````
10:29:19.223930 {0x00005650} [debug] network : clientnetworking.cpp:373 : Connected to server at host_name: 127.0.0.1  address: 127.0.0.1  port: 12346
10:29:19.223930 {0x00005650} [debug] network : clientnetworking.cpp:397 : Connecting to server took 503 ms.
10:29:19.223930 {0x00005650} [debug] network : clientnetworking.cpp:400 : ConnectToServer() : starting networking thread
10:29:19.224930 {0x00005650} [debug] client : humanclientapp.cpp:1113 : HumanClientApp::HandleFocusChange(Gained Focus)
10:29:19.440979 {0x0000656c} [debug] network : clientnetworking.cpp:512 : Client connection disconnected, due to connection reset from server.
10:29:19.533000 {0x00005650} [debug] client : humanclientapp.cpp:1559 : HumanClientApp::DisconnectedFromServer
10:29:19.533000 {0x00005650} [debug] client : humanclientapp.cpp:1449 : HumanClientApp::ResetToIntro
10:29:19.533000 {0x00005650} [debug] FSM : humanclientfsm.cpp:1260 : QuittingGame terminated server process.
10:29:19.533000 {0x00005650} [error] client : process.cpp:169 : Process::Impl::Kill : Error terminating process: Access is denied.
10:29:19.533000 {0x00005650} [debug] network : clientnetworking.cpp:459 : ClientNetworking::SetPlayerID: player id set to: -1
10:29:19.534000 {0x00005650} [debug] client : sitreppanel.cpp:768 : SitRepPanel::Update()
10:29:19.534000 {0x00005650} [debug] client : universe.cpp:220 : Reset id allocators with highest object id = -1 and highest design id = -1
10:29:22.158341 {0x00005650} [debug] client : humanclientapp.cpp:1449 : HumanClientApp::ExitApp
10:29:22.158341 {0x00005650} [debug] client : humanclientapp.cpp:376 : HumanClientApp exited cleanly.
10:29:22.171345 {0x00005650} [debug] client : universe.cpp:220 : Reset id allocators with highest object id = -1 and highest design id = -1
10:29:22.171345 {0x00005650} [debug] client : chmain.cpp:321 : Human client main exited cleanly.
````

**Comment by o01eg** [comment]
@geoffthemedio I've added additional logging. Could you run it again with it?

**Comment by geoffthemedio** [comment]
now server log ends with
```
11:59:24.201388 {0x00004e9c} [debug] server : commonframework.cpp:67 : Initializing C++ interfaces for Python
11:59:24.201388 {0x00004e9c} [debug] server : serverframework.cpp:60 : Initializing server Python modules
11:59:24.306411 {0x00004e9c} [info] log : logger.cpp:200 : Added logger named "python"
11:59:24.306411 {0x00004e9c} [info] log : logger.cpp:339 : Setting "python" logger threshold to "debug".
11:59:24.306411 {0x00004e9c} [debug] log : loggerwithoptionsdb.cpp:105 : Configure log source "python" from optionsDB using threshold debug
11:59:24.306411 {0x00004e9c} [debug] python : configure_logging.py:221 : Python stdout and stderr are redirected to ai process.
11:59:24.306411 {0x00004e9c} [info] python : configure_logging.py:242 : The python logger is initialized with a log level of DEBUG
11:59:24.315413 {0x00004e9c} [debug] server : enumwrapper.cpp:79 : WrapGameStateEnums: Wrap BuildType enum
11:59:24.315413 {0x00004e9c} [debug] server : enumwrapper.cpp:82 : WrapGameStateEnums: Wrap BuildType enum 0019E30C
````

**Comment by o01eg** [comment]
@geoffthemedio Wonder if it's issue with msvc2017, I've changed structure to static std::map.

**Comment by geoffthemedio** [comment]
That seems to work...
````
13:30:37.235113 {0x0000279c} [debug] python : configure_logging.py:221 : Python stdout and stderr are redirected to ai process.
13:30:37.235113 {0x0000279c} [info] python : configure_logging.py:242 : The python logger is initialized with a log level of DEBUG
13:30:37.244115 {0x0000279c} [debug] server : enumwrapper.cpp:79 : WrapGameStateEnums: Wrap BuildType enum
13:30:37.244115 {0x0000279c} [debug] server : enumwrapper.cpp:82 : WrapGameStateEnums: Wrap BuildType enum 014C2ED0
13:30:37.244115 {0x0000279c} [debug] server : enumwrapper.cpp:82 : WrapGameStateEnums: Wrap BuildType enum 014F4CA0
13:30:37.244115 {0x0000279c} [debug] server : enumwrapper.cpp:82 : WrapGameStateEnums: Wrap BuildType enum 014C2DB0
13:30:37.244115 {0x0000279c} [debug] server : enumwrapper.cpp:82 : WrapGameStateEnums: Wrap BuildType enum 014C2D90
13:30:37.244115 {0x0000279c} [debug] server : enumwrapper.cpp:82 : WrapGameStateEnums: Wrap BuildType enum 0152CAB0
13:30:37.244115 {0x0000279c} [debug] server : enumwrapper.cpp:82 : WrapGameStateEnums: Wrap BuildType enum 0152CB90
13:30:37.244115 {0x0000279c} [debug] server : enumwrapper.cpp:82 : WrapGameStateEnums: Wrap BuildType enum 0152CD50
13:30:37.244115 {0x0000279c} [debug] server : enumwrapper.cpp:85 : WrapGameStateEnums: BuildType enum wrapped
13:30:37.248116 {0x0000279c} [warn] python : auth.py:36 : Can't read auth file C:/Users/g_top/AppData/Roaming/FreeOrion/auth.txt: <class 'FileNotFoundError'> [Errno 2] No such file or directory: 'C:/Users/g_top/AppData/Roaming/FreeOrion/auth.txt'
13:30:37.248116 {0x0000279c} [info] python : auth.py:41 : Auth initialized
13:30:37.249116 {0x0000279c} [info] python : chat.py:14 : Chat initialized
````
(continues...)

**Comment by o01eg** [comment]
Nice, I wonder if `constexpr initializer_list` works with MSVC 2019.

**Comment by o01eg** [comment]
@Vezzra Could you check if it works under MacOS?

**Review by Vezzra — APPROVED** [review]
@o01eg, it does. I can build and run FO, start a quickstart game and do a few turns. No apparent issues, but I didn't test thoroughly of course. Just verified that I can start a game.

Can't comment on the code.

**Pull request merged**